lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_firmware.py

```diff
@@ -0,0 +1,228 @@
+# Other trademarks may be trademarks of their respective owners.
+#
+
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = """
+---
+module: dellemc_idrac_firmware
+short_description: Firmware update from a repository on a network share (CIFS, NFS).
+version_added: "2.8"
+description:
+    - Update the Firmware by connecting to a network share (either CIFS or NFS) that contains a catalog of
+        available updates.
+    - Network share should contain a valid repository of Update Packages (DUPs) and a catalog file describing the DUPs.
+    - All applicable updates contained in the repository are applied to the system.
+    - This feature is available only with iDRAC Enterprise License.
+options:
+    idrac_ip:
+        required: True
+        description: iDRAC IP Address.
+    idrac_user:
+        required: True
+        description: iDRAC username.
+    idrac_pwd:
+        required: True
+        description: iDRAC user password.
+    idrac_port:
+        required: False
+        description: iDRAC port.
+        default: 443
+    share_name:
+        required: True
+        description: CIFS or NFS Network share.
+    share_user:
+        required: False
+        description: Network share user in the format 'user@domain' or 'domain\\user' if user is
+            part of a domain else 'user'. This option is mandatory for CIFS Network Share.
+    share_pwd:
+        required: False
+        description: Network share user password. This option is mandatory for CIFS Network Share.
+    share_mnt:
+        required: True
+        description: Local mount path of the network share with read-write permission for ansible user.
+            This option is mandatory for Network Share.
+    reboot:
+        required: False
+        description: Whether to reboots after applying the updates or not.
+        default: False
+        type: bool
+    job_wait:
+        required:  True
+        description: Whether to wait for job completion or not.
+        type: bool
+        default: True
+    catalog_file_name:
+        required: False
+        description: Catalog file name relative to the I(share_name).
+        type: str
+        default: 'Catalog.xml'
+
+requirements:
+    - "omsdk"
+    - "python >= 2.7.5"
+author: "Rajeev Arakkal (@rajeevarakkal)"
+"""
+
+EXAMPLES = """
+---
+- name: Update firmware from repository on a Network Share
+  dellemc_idrac_firmware:
+       idrac_ip:   "{{ idrac_ip }}"
+       idrac_user: "{{ idrac_user }}"
+       idrac_pwd:  "{{ idrac_pwd }}"
+       share_name: "xx.xx.xx.xx:/share"
+       share_user: "{{ share_user }}"
+       share_pwd:  "{{ share_pwd }}"
+       share_mnt: "/mnt/share"
+       reboot:     True
+       job_wait:   True
+       catalog_file_name:  "Catalog.xml"
+"""
+
+RETURN = """
+---
+msg:
```

Comment:
With all I saw I think that you need to extend your schema:
1) have a string `msg` key, because I saw you putting strings instead of dict there sometimes
2) some structure for extra info: either one dict with all data or a set of separate scalar keys (which is probably easier to manage)

Think of it as a public API: whoever will run your module will register the result and this whole structure will be available to them to use/inspect.
Try to think of the most useful way to reuse this data you return.

Reply:
updated the return type to `string`

Reply:
msg should be smth added by you, while there should also be a dict with server response separately.

Reply:
Then, this return schema should be updated. It's currently completely different from what you describe. I guess it should be:
```yaml
---
msg:
  type: string
  ...
  returned: always
  sample: <actual string message>
  ...
update_status:
  type: dict
  ...
  returned: success
  sample:
    CompletionTime: ...
    Id: ...
    ...
```

Reply:
(return value is a _nested_ structure, so you don't have to try putting everything in one key: different data should go to different keys)